### ISSUE TITLE
user.Username vs user.Id.OpaqueId

### ISSUE THREAD

**Issue by michielbdejong** [cs3org/reva#2377]
Authenticate [returns](https://github.com/cs3org/cs3apis/blob/e3ac6ae/cs3/auth/provider/v1beta1/provider_api.proto#L77) a user object which has a [`username` string](https://github.com/cs3org/cs3apis/blob/e3ac6ae9a66f8d7aed0662a28fb46ebdd012d356/cs3/identity/user/v1beta1/resources.proto#L50) but also an [`Id` UserId](https://github.com/cs3org/cs3apis/blob/e3ac6ae9a66f8d7aed0662a28fb46ebdd012d356/cs3/identity/user/v1beta1/resources.proto#L49) which contains the usual `idp` and `opaque_id`.

This user is then [set into the context of the CreateHome request](https://github.com/cs3org/reva/blob/v1.17.0/internal/grpc/services/gateway/authprovider.go#L153) by the gateway service auth provider.

When handling the CreateHome request in the storage fs implementation, I was reading the `user.Username` from there, but maybe it's better to read `user.Id.OpaqueId` instead?

Is there a situation where the two would be different? Maybe when there are multiple IDP's involved in a single site?

**Comment by michielbdejong** [comment]
fwiw, caching in the authprovider is done on the basis of `res.User.Id.OpaqueId`

**Comment by butonic** [comment]
IMNSHU we need to be able to distinguish users by id. There might actually be two user with the same name when multiple idbs are involved, which is why the the code should never rely on the username. That the home path (and other) layouts can be configured to use the username is a huge shotgun pointed at the admins foot. That is the reason why ocis defaults to `/users/{.Id.OpaqueId}`.

**Comment by michielbdejong** [comment]
> That the home path (and other) layouts can be configured to use the username is a huge shotgun pointed at the admins foot. 

Good point! Probably the only thing that would be really really safe would be something like `/users/{.Id.Idp}/{.Id.OpaqueId}`

I guess the `.Id.OpaqueId` is provided by the IDP in question, and the `Username` is ultimately provided by the auth manager.

In our case the auth manager has only one IDP backend, so then it's simple, we just set `.Id.Idp` to a fixed string like `'local'` and then just copy `.Id.OpaqueId` to `.Username`. but it's definitely something to be aware of for sites that admit multiple IDPs
